turbopack/crates/turbo-tasks-backend/src/lib.rs

```diff
@@ -46,30 +47,39 @@ pub fn lmdb_backing_storage(
         startup_cache::StartupCacheLayer,
     };
 
-    let path = handle_db_versioning(path, version_info)?;
-    let fresh_db = is_fresh(&path);
-    let database = crate::database::lmdb::LmbdKeyValueDatabase::new(&path)?;
+    check_db_invalidation_and_cleanup(base_path)?;
+    let versioned_path = handle_db_versioning(base_path, version_info)?;
+    let fresh_db = is_fresh(&versioned_path);
+    let database = crate::database::lmdb::LmbdKeyValueDatabase::new(&versioned_path)?;
     let database = FreshDbOptimization::new(database, fresh_db);
-    let database = StartupCacheLayer::new(database, path.join("startup.cache"), fresh_db)?;
+    let database =
+        StartupCacheLayer::new(database, versioned_path.join("startup.cache"), fresh_db)?;
     let database = ReadTransactionCache::new(database);
-    Ok(KeyValueDatabaseBackingStorage::new(database))
+    Ok(KeyValueDatabaseBackingStorage::new(
+        database,
+        Some(base_path.to_owned()),
+    ))
 }
 
 pub type TurboBackingStorage = KeyValueDatabaseBackingStorage<TurboKeyValueDatabase>;
 
 pub fn turbo_backing_storage(
-    path: &Path,
+    base_path: &Path,
     version_info: &GitVersionInfo,
 ) -> Result<TurboBackingStorage> {
-    let path = handle_db_versioning(path, version_info)?;
-    let database = TurboKeyValueDatabase::new(path)?;
-    Ok(KeyValueDatabaseBackingStorage::new(database))
+    check_db_invalidation_and_cleanup(base_path)?;
```

Comment:
should this be part of handle_db_versioning?  seems somewhat related (we could think of this as some kind of 'versioning poison pill'?  that might make it a bit harder to forget if a new backingstorage is added

Reply:
Okay, but I'm going to change the name of the `new` method to something like `initialize` in that case. I don't like constructors that do I/O, as it's often assumed that constructors are cheap.

Reply:
maybe `open` is a better name?

Reply:
Oh, I see why I originally did this: The database clearing must happen before LMDB's fresh DB check and startup cache steps. It's also more ideal, but not strictly necessary for it to happen before the versioning check (because we invalidate all versions). @sokra

Reply:
I've split the constructor for `KeyValueDatabaseBackingStorage` into two: one for versioned on-disk databases and one for in-memory databases (which don't need cache clearing).

This means that I also moved the versioning logic into the `KeyValueDatabaseBackingStorage` constructor, as it seems like we always want that for on-disk databases.